Adds the _cluster object to the AsyncStatusResponse XContent when Clusters has a non-zero count total

quux00 opened the pull request:
During a cross cluster search, the `_cluster` object in the SearchResponse will have a non-zero `total` count and thus is useful status information for the user. That information is included in the `_async_search/<id>` endpoint already because it is part of the underlying SearchResponse.

However, that information is not present in the `_async_search/status/<id>` endpoint, since that does not show the SearchResponse details. With a cross-cluster search this is useful information to understand (1) how many clusters are being searched; (2) whether they are successful or skipped, so it is being added in this commit.

When an async search is performed against the local cluster only , the `_cluster` object will not be present in the AsyncStatusResponse, since it is not informative (TransportSearchAction uses Clusters.EMPTY for executeLocalSearch).

quux00 committed:
Adds the _cluster object to the AsyncStatusResponse XContent when Clusters has a non-zero count total.

When CCS is done, the _cluster object will have a non-zero `total` count and thus
is useful status information for the user.

When a local cluster only async search is done, the _cluster object will not be present in the AsyncStatusResponse.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

quux00 commented:
Manual testing examples:

Local cluster only searches do not show the clusters _object from either from the `_async_search` or the `_async_search/status` endpoints:

```
POST blogs/_async_search?ccs_minimize_roundtrips=true
// response
{                                            
  "id": "Fjl2N3plbHRMVDl5Q29KamF6N3NVZW      cbTXBCOEliLUdSdzZSZmZkWW15ZmdNQToyMDAx",
  "is_partial": true,                        
  "is_running": true,                        
  "start_time_in_millis": 1685725885651      ,
  "expiration_time_in_millis": 16861578      85651,
  "response": {                              
    "took": 1028,                            
    "timed_out": false,                      
    "terminated_early": false,               
    "num_reduce_phases": 0,                  
    "_shards": {                             
      "total": 3,                            
      "successful": 0,                       
      "skipped": 0,                          
      "failed": 0                            
    },                                       
    "hits": {                                
      "total": {                             
        "value": 0,                          
        "relation": "gte"                    
      },                                     
      "max_score": null,                     
      "hits": []                             
    }                                        
  }                                          
}                                            
                                             
// intermediate (while still running)                          
GET /_async_search/status/Fjl2N3plbHRMVDl5Q29KamF6N3NVZWcbTXBCOEliLUdSdzZSZmZkWW15ZmdNQToyMDAx
{                                            
  "id": "Fk1wakRkLUd3UzFDWUxpcDc0eGtLMmcbTXBCOEliLUdSdzZSZmZkWW15ZmdNQToyMTEx",
  "is_running": true,                        
  "is_partial": true,                        
  "start_time_in_millis": 1685725966337      ,
  "expiration_time_in_millis": 16861579      66337,
  "_shards": {                               
    "total": 3,                              
    "successful": 0,                         
    "skipped": 0,                            
    "failed": 0                              
  }                                          
}                                            
                                             
                                             
// final                                    
GET /_async_search/status/Fjl2N3plbHRMVDl5Q29KamF6N3NVZWcbTXBCOEliLUdSdzZSZmZkWW15ZmdNQToyMDAx
                                             
{                                            
  "id": "Fjl2N3plbHRMVDl5Q29KamF6N3NVZWcbTXBCOEliLUdSdzZSZmZkWW15ZmdNQToyMDAx",
  "is_running": false,                       
  "is_partial": false,                       
  "start_time_in_millis": 1685725885651      ,
  "expiration_time_in_millis": 1686157885651,
  "_shards": {                               
    "total": 3,                              
    "successful": 3,                         
    "skipped": 0,                            
    "failed": 0                              
  },                                         
  "completion_status": 200                   
}                                            
```

whereas searches with one or more remote clusters (CCS) shows the _cluster object from both endpoints:

```
POST remote1:*,remote2:*/_async_search?ccs_minimize_roundtrips=true
                                           
{                                          
  "id": "FlRYVE81OWZDUm8tMmN3RjQtNEw3SEEbTXCOEliLUdSdzZSZmZkWW15ZmdNQToyMjMz",
  "is_partial": true,                      
  "is_running": true,                      
  "start_time_in_millis": 1685726121946,   
  "expiration_time_in_millis": 168615812194,
  "response": {                            
    "took": 1005,                          
    "timed_out": false,                    
    "terminated_early": false,             
    "num_reduce_phases": 0,                
    "_shards": {                           
      "total": 0,                          
      "successful": 0,                     
      "skipped": 0,                        
      "failed": 0                          
    },                                     
    "_clusters": {                         
      "total": 2,                          
      "successful": 0,                     
      "skipped": 0                         
    },                                     
    "hits": {                              
      "total": {                           
        "value": 0,                        
        "relation": "gte"                  
      },                                   
      "max_score": null,                   
      "hits": []                           
    }                                      
  }                                        
}                                          
                                           
// intmd                                   
GET /_async_search/status/FlRYVE81OWZDUm8tMmN3RjQtNEw3SEEbTXCOEliLUdSdzZSZmZkWW15ZmdNQToyMjMz
{                                          
  "id": "FlRYVE81OWZDUm8tMmN3RjQtNEw3SEEbTXCOEliLUdSdzZSZmZkWW15ZmdNQToyMjMz",
  "is_running": true,                        
  "is_partial": true,                        
  "start_time_in_millis": 1685726220696,     
  "expiration_time_in_millis": 168615822069  ,
  "_shards": {                               
    "total": 0,                              
    "successful": 0,                         
    "skipped": 0,                            
    "failed": 0                              
  },                                         
  "_clusters": {                             
    "total": 2,                              
    "successful": 0,                         
    "skipped": 0                             
  }                                          
}                                            
                                             
// final                                     
GET /_async_search/status/FlRYVE81OWZDUm8tMmN3RjQtNEw3SEEbTXCOEliLUdSdzZSZmZkWW15ZmdNQToyMjMz
{                                            
  "id": "FlRYVE81OWZDUm8tMmN3RjQtNEw3SEEbTXCOEliLUdSdzZSZmZkWW15ZmdNQToyMjMz",
  "is_running": false,                       
  "is_partial": false,                       
  "start_time_in_millis": 1685726121946,     
  "expiration_time_in_millis": 168615812194  ,
  "_shards": {                               
    "total": 6,                              
    "successful": 6,                         
    "skipped": 0,                            
    "failed": 0                              
  },                                         
  "_clusters": {                             
    "total": 2,                              
    "successful": 2,                         
    "skipped": 0                             
  },                                         
  "completion_status": 200                   
}                                          
```

javanna commented on x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java:
```diff
@@ -241,6 +241,11 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
      */
     synchronized AsyncStatusResponse toStatusResponse(String asyncExecutionId, long startTime, long expirationTime) {
+        SearchResponse.Clusters clustersInStatus = null;
+        if (clusters != null && clusters.getTotal() > 0) {
```
Can we have a clusters instance that has total == 0 in practice?

quux00 replied:
Yes, TransportSearchAction uses Clusters.EMPTY for [executeLocalSearch](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L313). 

That is why for searches against local clusters only we don't display the `_clusters` object. It's content-free for that scenario.

The CCS pathway (even with a local cluster) doesn't use that pathway in TransportSearchAction. Instead, it sets up an informative Clusters object.

javanna replied:
good point. I forgot.

javanna commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncStatusResponse.java:
```diff
@@ -36,6 +36,12 @@ public class AsyncStatusResponse extends ActionResponse implements SearchStatusR
+    // This only needs to be handled on the coordinator, so this instance is transient (not Writeable).
+    // The SearchResponse has a Writeable copy that gets written to the async-search
+    // system index, so we can retrieve it from there after the task is finished
+    private transient SearchResponse.Clusters clusters;
```
what happens if node1 is the coordinator for the ongoing search, and you ask node2 for the status of that same async search?

quux00 replied:
Oh, thanks for that. I'm not sure how that works. Would the node 2 coordinator be contacting node 1 to get the in-memory async search status?

So in that case we would need to have Clusters added to the transport serialization so that node 1 can send it to node 2. Is that right?

quux00 replied:
Yeah, I set up a 2-node cluster and tested this case and if you go to node2 for status you don't get the `_clusters` entry, so I'll need to make that field Writeable and do a Transport Version bump in this PR.

javanna reviewed (commented):
left a couple of questions

quux00 committed:
SearchResponse.Clusters is now Writeable in the AsyncStatusResponse object so it can be obtained

by get-async-search-status coordinators that are not the coordinator for the async search itself.
This requires a Transport Version bump.

quux00 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncStatusResponse.java:
```diff
@@ -145,9 +150,13 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeVInt(failedShards);
-        if (isRunning == false) {
+        if (isRunning == false) { /// MP: Hmm, is this going to be a problem? Why is this not writeOptionalString?
             RestStatus.writeTo(out, completionStatus);
```
Having a variably written (meaning sometimes yes, sometimes no) field here now in the middle of the serialization seems problematic. I'm not sure why this was done this way instead of using `writeOptionalString`.

Is this a concern now?
Should I have change that to a writeOptionalString string in my `TransportVersion.V_8_500_008` if clause for safety?

In that case the code would be:

```
        if (out.getTransportVersion().before(TransportVersion.V_8_500_008)) {
            if (isRunning == false) { 
                RestStatus.writeTo(out, completionStatus);
            }
        } else {
            if (isRunning == false && completionStatus != null) {
                out.writeOptionalString(completionStatus.name());
            } else {
                out.writeOptionalString(null);
            }
            // optional since only CCS uses is; it is null for local-only searches
            out.writeOptionalWriteable(clusters);
        }
```

javanna replied:
writeOptionalBoolean writes a fixed boolean that signals whether something else follows or not. In this case, isRunning is serialized up above and drives whether the completion status is written or not. I see how this may not be super readable, but I don't think it causes issues, technically, unless you see that the condition for writing the completion status should change to not depend on isRunning alone.

quux00 replied:
OK. Since `isRunning=false` is a terminal state, I guess this is safe. It seems odd to me however. The "optional" part of the serialization mechanism allows for nulls, so why not encode the null in the Writeable input/output when isRunning=true? The underlying mechanism already handles this case, so adding ad-hoc logic on top of it seems redundant and confusing (like there is some reason it can't be encoded as a null in the serialization system).

Since it is already there, I'm fine with leaving it if that's the path of least resistance. I just wanted to see if you though this is something that should be changed.

javanna replied:
I see what you mean, yet the boolean is already serialized, only a few lines above, that is my line of thinking.

quux00 committed:
Adjusted AsyncStatusResponse to check for failures before searchResponse when looking to set completion status.

Added several new unit tests in AsyncStatusResponseTests

javanna commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncStatusResponse.java:
```diff
@@ -145,9 +150,13 @@ public void writeTo(StreamOutput out) throws IOException {
         }
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_500_008)) {
+            // optional since only CCS uses is; it is null for local-only searches
+            out.writeOptionalWriteable(clusters);
```
Once we add new fields to Clusters, we will need to handle bwc within Clusters and all the classes that include a Clusters instance and serialize it will inherit that in their serialization logic? What would VersionedNamedWriteable help with in this case?

quux00 replied:
OK, sounds good. Thanks for confirming this is the right approach then.

javanna commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncStatusResponse.java:
```diff
@@ -132,9 +150,13 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeVInt(skippedShards);
         out.writeVInt(failedShards);
-        if (isRunning == false) {
+        if (isRunning == false) { /// MP: Hmm, is this going to be a problem? Why is this not writeOptionalString?
```
just checking that you don't forget to remove this comment ;)

javanna commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/search/action/AsyncStatusResponse.java:
```diff
@@ -132,9 +150,13 @@ public void writeTo(StreamOutput out) throws IOException {
             RestStatus.writeTo(out, completionStatus);
         }
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_500_009)) {
+            // optional since only CCS uses is; it is null for local-only searches
```
This needs to be nullable because older nodes won't send it over the wire. I am less clear on having it null for local-only searches. Shouldn't we use EMPTY in this case like we normally do?

quux00 replied:
Currently (before this ticket) `_clusters` is not shown for local-only searches in the async-status/status response, so that has been the default already.

And in addition, I'm not sure why we use EMPTY for the normal non-CCS case, as that seems wrong to me. 

This type of output would be confusing to end users:
```
{
  "id": "FnhKR2FES241UmNtSjhGWUVTYkhGWWcaemJJaENTNW9SZFdaWnNmRUNzbE1VUTozNjU",
  "is_running": false,
  "is_partial": false,
  "start_time_in_millis": 1686227754043,
  "expiration_time_in_millis": 1686659754043,
  "_shards": {
    "total": 3,
    "successful": 3,
    "skipped": 0,
    "failed": 0
  },
  "_clusters": {
    "total": 0,  // why is this zero?
    "successful": 0, // didn't my local cluster search work?
    "skipped": 0
  },
  "completion_status": 200
}
```

IMO, the standard non-CCS default Clusters object should be `Clusters{total=1, successful=1, skipped=0}`. Using empty implies that we don't want to count the local cluster as a cluster. But if that's true, then should we also not count it for CCS? But then that would be also confusing to end users.

If we add EMPTY here for the local-only, we could choose to not serialize it, but that requires extra logic in the code above or if we serialize it but don't want to show it, then that requires extra logic in the toXContent methods. So I chose the path of least resistance which is to just have it be `null`, which matches current functionality for local-only searches.

If you want EMPTY serialized, then I would say we need to add extract logic in `toXContent` to not output it to the REST response.

javanna reviewed (approved):
left a question, LGTM otherwise

quux00 committed:
Removed debugging comment
